gphoto2: Add option for group name

l0b0 opened the pull request:
## Description of changes

Enables referencing it in configuration dynamically, rather than with a magic string.

Closes #294871.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bjornfor commented on nixos/modules/programs/gphoto2.nix:
```diff
@@ -11,12 +14,20 @@ with lib;
+          Group created when gphoto2 is enabled.
+
+          To grant digital camera access to a user, the user must be part of this group:
+          `users.users.alice.extraGroups = [options.programs.gphoto2.group.value];`.
```
Shouldn't this be `config.programs...` instead of `options.programs...`? (I think `options...` is only the option defaults, whereas `config...` reflects the fully merged user config.)

l0b0 replied:
I think you're right. Like this?

bjornfor commented on pkgs/development/libraries/libgphoto2/default.nix:
```diff
@@ -14,7 +15,10 @@
-
+let
+  inherit (nixos {}) config;
+  group = config.programs.gphoto2.group;
```
Won't this reference the default value and not the one from the users' nixos config? Maybe it's better to make the option read-only (I think that's possible), so that at least the magic string goes away?

SuperSandro2000 commented on pkgs/development/libraries/libgphoto2/default.nix:
```diff
@@ -14,7 +15,10 @@
 }:
-
+let
+  inherit (nixos {}) options;
+  group = options.programs.gphoto2.group.value;
```
I am pretty sure that this won't work on none nixos systems. We should probably just make this a string.

l0b0 replied:
How's the current solution?

SuperSandro2000 commented on nixos/modules/programs/gphoto2.nix:
```diff
@@ -11,12 +14,20 @@ with lib;
+
+          To grant digital camera access to a user, the user must be part of this group:
+          `users.users.alice.extraGroups = [config.programs.gphoto2.group];`.
+
+          By default, a group named `${defaultGroup}` will be created.
```
```suggestion
```
the default value of the option is already documented in all docs.

Also then we can drop the let in entirely.

SuperSandro2000 commented on nixos/tests/gphoto2/custom-group-name.nix:
```diff
@@ -0,0 +1,22 @@
+  testScript = ''
+    ${machineName}.start()
+
+    ${machineName}.succeed("gphoto2 --version")
+    ${machineName}.succeed("getent group my-group")
+
+    # Make sure it doesn't also create the default group
+    ${machineName}.fail("getent group camera")
```
Running the program with --version is supposed to be done in passthru with the helper. Checking if the group is created doesn't add much value especially if we need to start an entire VM for that. We should do a proper integration test if the program works as intended or drop this.

l0b0 replied:
Wouldn't that require simulating an actual camera device? That sounds really difficult. The value of the current test is that it does verify that the change does what it says on the tin.

SuperSandro2000 commented on nixos/tests/gphoto2/defaults.nix:
```diff
@@ -0,0 +1,16 @@
+    ${machineName}.start()
+
+    ${machineName}.succeed("gphoto2 --version")
+    ${machineName}.succeed("getent group camera")
```
Also I am skeptical about this test. We don't even check if the program is using the group correctly. So basically the same as above.

l0b0 committed:
gphoto2: Format files using nixfmt-rfc-style

l0b0 committed:
gphoto2: Add option for group name

Enables referencing it in configuration dynamically, rather than with a
magic string.

Closes #294871.

bjornfor commented on pkgs/by-name/li/libgphoto2/package.nix:
```diff
@@ -1,20 +1,25 @@
-
+let
+  inherit (nixos { }) config;
+  group = config.programs.gphoto2.group or "camera";
```
I don't see how this can work. Isn't this configuring a package build based on a static NixOS default value?

l0b0 replied:
If the user sets `programs.gphoto2.group = "foo"`, then the gphoto2 group will be "foo". If they don't set anything then it will be defaulted to "camera". Is that not what we want?

SuperSandro2000 replied:
The option already has a fallback but I thought we need this when where are not on a nixos system and just use nixpkgs. Not sure if we then eval the module system just for that.

bjornfor replied:
> If the user sets programs.gphoto2.group = "foo", then the gphoto2 group will be "foo".

Have you tested it?

l0b0 replied:
If you have some information, why not share it? I tried testing it, and first my entire computer crashed, and then there seems to be some unrelated issue with systemd on master.

bjornfor replied:
> If you have some information, why not share it?

I don't have any more information than is in this thread: I think this doesn't work because I think `nixos {} config` returns static defaults instead of taking into account the users config.

bjornfor replied:
Btw, you can probably test/apply this PR onto whatever nixpkgs revision your NixOS system is built from -- it doesn't have to be master.

l0b0 replied:
> I don't have any more information than is in this thread: I think this doesn't work because I think nixos {} config returns static defaults instead of taking into account the users config.

OIC, would absolutely be good to test this. 

> Btw, you can probably test/apply this PR onto whatever nixpkgs revision your NixOS system is built from -- it doesn't have to be master.

I'm afraid I don't know how to do that. The official [Nixpkgs/Reviewing changes](https://wiki.nixos.org/wiki/Nixpkgs/Reviewing_changes#Flakes) article doesn't seem to explain how to do that.

l0b0 replied:
Hmh, writing a test it turns out I don't know how to get the current `config.programs.gphoto2.group` value in `pkgs/by-name/li/libgphoto2/package.nix`. I've tried a few variants, but I don't understand how other packages are getting [things like `config.pulseaudio`](https://github.com/l0b0/nixpkgs/blob/gphoto2-group-name-config/pkgs/by-name/ae/aegisub/package.nix#L37-L37) seemingly without descending into `config.services` first.

bjornfor replied:
1. I think it's not possible. The only way would be to override the package inside a NixOS module.
2. config.pulseaudio is nixpkgs configuration, not NixOS.

bjornfor replied:
> I'm afraid I don't know how to do that.

Take the nixpkgs git working tree you already have (where this PR is made from), checkout a new branch from the nixpkgs revision your NixOS is built from, git cherry-pick the PR branch, run "NIX_PATH=nixpkgs=/path/to/my/nixpkgs nixos-rebuild ...".

l0b0 replied:
Thanks for the patience! I've changed this around to override the libgphoto2 group when enabling `programs.gphoto2`. What do you think?

bjornfor replied:
I think that will work.

However, I'm not so happy about having to rebuild software just to change group config. In your PR message you say the point is to not use magic string in config, but that can be solved by hardcoding the group in the package and making a read-only NixOS option that propagates that value.

Alternatively, if you really want to be able to change the group in NixOS config, I think the option docstring should warn about the rebuild cost.

Note: this is not me blocking the PR, but I'm not comfortable approving it either. If anyone else think it's fine, they can merge.

l0b0 replied:
Good points! I'll go for a read-only option instead. The main point was to be able to refer to the group in things like `users.users.<name>.extraGroups`, after all. How should I expose the group in libgphoto2? I can't seem to find any examples of arbitrary attributes like this in other packages.

l0b0 replied:
I've gone for `passthru.group` and a read-only option, so there should be no more rebuilds.

l0b0 replied:
[That doesn't work](https://github.com/NixOS/nixpkgs/actions/runs/14155119464/job/39653362389?pr=296832). Not sure how such a reference is meant to be done.

bjornfor replied:
About the CI error, try adding defaultText next to default. That should avoid the dependency it complains about.

l0b0 replied:
OK, let's see how this works.

bjornfor reviewed (dismissed):
My concern: https://github.com/NixOS/nixpkgs/pull/296832/files#r1986152508
